lib/wasi/src/runtime/task_manager/mod.rs

```diff
@@ -2,43 +2,125 @@
 #[cfg(feature = "sys-thread")]
 pub mod tokio;
 
+use std::task::{Context, Poll};
 use std::{pin::Pin, time::Duration};
 
 use ::tokio::runtime::Handle;
 use futures::Future;
-use wasmer::{Memory, MemoryType, Module, Store, StoreMut};
+use wasmer::{AsStoreMut, AsStoreRef, Memory, MemoryType, Module, Store, StoreMut, StoreRef};
+use wasmer_wasix_types::wasi::{Errno, ExitCode};
 
 use crate::os::task::thread::WasiThreadError;
+use crate::syscalls::AsyncifyFuture;
+use crate::{capture_snapshot, InstanceSnapshot, WasiEnv, WasiFunctionEnv, WasiThread};
 
 #[derive(Debug)]
-pub struct SpawnedMemory {
-    pub ty: MemoryType,
+pub enum SpawnMemoryType<'a> {
+    CreateMemory,
+    CreateMemoryOfType(MemoryType),
+    ShareMemory(Memory, StoreRef<'a>),
+    CopyMemory(Memory, StoreRef<'a>),
```

Comment:
I don't love that we are storing `StoreRef` which is pretty much internal to the VM. This might break JSC again.
Are there other ways to achieve this?

Reply:
unsafe ones yes... the problem is that memory is a pointer to something that exists on in a single thread when running in the browser... the alternative is to change this to VMMemory

Reply:
`VMMemory` alternative will be non viable for JSC. Do you know if having a `Box<dyn AsStoreRef<'a>>` would work here @john-sharratt ?

Reply:
Putting it on the heap won't change that fact its still a lifetime limited memory reference - if you want a heap reference we would need to do something like `Arc<Mutex<Store>>` but that will be a mess

Reply:
For now this will have to do, left a TODO to look at it again